lib/pac55xx/can.c

```diff
@@ -0,0 +1,470 @@
+	/* set mask */
+	word = ((~id_mask << 3) & CAN_ACR_SINGLE_EXT_ID) | CAN_ACR_SINGLE_EXT_RTR;
+	CAN_AMR(canport) = __builtin_bswap32(word);
+	/* 1: single filter */
+	CAN_ISR_SR_CMR_MR_SET(canport, CAN_MR_AFM);
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief CAN Filter Single Extended Frame w/RTR set
+Note: Acceptance Code Mask Register values of 1 indicate the filter is to
+ignore the bit. However, standard CAN driver APIs use a positive logic for the
+mask. So this function inverts the mask to make this more portable/intuitive.
+
+@param[in] canport Unsigned int32. CAN block register base address.
+@param[in] id Unsigned int32. CAN ID. Only bits 28:0 are used.
+@param[in] id_mask Unsigned int32. CAN ID mask. Only bits 28:0 are used.
+@param[in] rtr bool. Remote Request bit value.
+@param[in] rtr_mask bool. Remote Request mask value.
+*/
+void can_filter_single_ext_rtr(uint32_t canport, uint32_t id, uint32_t id_mask) {
+	/* set value */
+	uint32_t word = ((id << 3) & CAN_ACR_SINGLE_EXT_ID) | CAN_ACR_SINGLE_EXT_RTR;
+	CAN_ACR(canport) = __builtin_bswap32(word);
+	/* set mask */
+	word = ((~id_mask << 3) & CAN_ACR_SINGLE_EXT_ID);
+	CAN_AMR(canport) = __builtin_bswap32(word);
+	/* 1: single filter */
+	CAN_ISR_SR_CMR_MR_SET(canport, CAN_MR_AFM);
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief CAN Enable IRQ
+@param[in] canport Unsigned int32. CAN block register base address.
+@param[in] irq Unsigned int8. IRQ bit(s).
+*/
+void can_enable_irq(uint32_t canport, uint8_t imr) {
+	/* set to 1 (not masked) to enable */
+	CAN_BTR1_BTR0_RMC_IMR(canport) |= (uint32_t)imr;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief CAN Disable IRQ
+@param[in] canport Unsigned int32. CAN block register base address.
+@param[in] irq Unsigned int8. IRQ bit(s).
+*/
+void can_disable_irq(uint32_t canport, uint8_t imr) {
+	/* set to 0 (masked) to disable */
+	CAN_BTR1_BTR0_RMC_IMR(canport) &= ~(uint32_t)imr;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief CAN Transmit Standard Frame
+@param[in] canport Unsigned int32. CAN block register base address.
+@param[in] id Unsigned int32. Message ID bits 10:0 used.
+@param[in] rtr bool. Remote Request bit value.
+@param[in] length Unsigned int8. Message payload length.
+@param[in] data Unsigned int8[]. Message payload data.
+@returns int 0 on success, or -1 on failure.
+*/
+int can_transmit_std(uint32_t canport, uint32_t id, bool rtr, uint8_t length,
+				const uint8_t *data) {
+	if (length > 8) {
+		return -1;
+	}
+
+	/* if TBS is 0, then not ready to transmit */
+	if ((CAN_ISR_SR_CMR_MR(canport) & CAN_SR_TBS) == 0) {
+		return -1;
+	}
+	uint32_t word = (length) | (rtr ? BIT6 : 0) | /* DLC/RTR/FF ==> 7:0 */
+					((id & 0x07F8) << 5) | /* ID 10:3 ==> 15:8 */
+					((id & 0x07) << 21)  | /* ID 2:0 ==> 23:21 */
```

Comment:
given the absolute _mountain_ of defines you added, how come these magic numbers are all still here?

Reply:
(sorry about the _mountain_ comment, I'm mixing some of this up with the fdcan PR for stm32, still, these sort of special magic should be defined somewhere.)

Reply:
no problem.  yeah, I wish there was a way to do a macro like BITS(2,5), which would result in 0x3C (all bits from 2 to 5 set).  There are so many of these magic numbers throughout this file...  There's got to be a better way!

Reply:
Added another hill of defines in the header to have a place for the magic numbers.